packages/ckeditor5-engine/src/model/utils/deletecontent.js

```diff
@@ -97,89 +97,254 @@ export default function deleteContent( model, selection, options = {} ) {
+	// <a><b>x[</b></a><c><d>]y</d></c>
+	// will become:
+	// <a>[</a><c>]<d>xy</d></c>
+	startPosition = writer.createPositionAfter( startElement );
+	endPosition = writer.createPositionBefore( endElement );
+
+	if ( !endPosition.isEqual( startPosition ) ) {
+		// In this case, before we merge, we need to move `startElement` to the `endPosition`:
+		// <a><b>x[</b></a><c><d>]y</d></c>
+		// becomes:
+		// <a>[</a><c><b>x</b>]<d>y</d></c>
+		writer.insert( startElement, endPosition );
+	}
+
+	// Remove empty start ancestors:
+	// <x><a><b>x[</b></a></x><c><d>]y</d></c>
+	// becomes:
+	// <x><a>[</a></x><c><b>x</b>]<d>y</d></c>
+	// So we can remove <a> and <x>.
+	while ( startPosition.parent.isEmpty ) {
+		const parentToRemove = startPosition.parent;
+
+		startPosition = writer.createPositionBefore( parentToRemove );
+
+		writer.remove( parentToRemove );
+	}
+
+	// Update endPosition after inserting and removing elements.
+	endPosition = writer.createPositionBefore( endElement );
+
+	// Merge two siblings:
+	// <a>x</a>[]<b>y</b> -> <b>xy</b>
+	mergeRight( writer, endPosition );
+
+	// Verify if there is a need and possibility to merge next level.
+	if ( !checkShouldMerge( writer.model.schema, startPosition, endPosition ) ) {
+		return;
+	}
+
+	// Continue merging next level.
+	mergeBranchesRight( writer, startPosition, endPosition, commonAncestor );
+}
+
+// There is no right merge operation so we need to simulate it.
+function mergeRight( writer, position ) {
+	const startElement = position.nodeBefore;
+	const endElement = position.nodeAfter;
+
+	if ( startElement.name != endElement.name ) {
+		writer.rename( startElement, endElement.name );
+	}
+
+	writer.clearAttributes( startElement );
+	writer.setAttributes( Object.fromEntries( endElement.getAttributes() ), startElement );
+
+	writer.merge( position );
+}
+
+// Verifies if merging is needed and possible.
+function checkShouldMerge( schema, startPosition, endPosition ) {
+	const startElement = startPosition.parent;
+	const endElement = endPosition.parent;
+
+	// If both positions ended up in the same parent, then there's nothing more to merge:
+	// <$root><p>x[</p><p>]y</p></$root> => <$root><p>xy</p>[]</$root>
+	if ( startElement == endElement ) {
+		return false;
+	}
+
+	// If one of the positions is a limit element, then there's nothing to merge because we don't want to cross the limit boundaries.
+	if ( schema.isLimit( startElement ) || schema.isLimit( endElement ) ) {
+		return false;
+	}
+
+	// Check if operations we'll need to do won't need to cross object or limit boundaries.
+	// E.g., we can't merge endElement into startElement in this case:
+	// <limit><startElement>x[</startElement></limit><endElement>]</endElement>
+	return checkCanBeMerged( startPosition, endPosition, schema );
+}
+
+// Returns the elements that are the ancestors of the provided positions that are direct children of the common ancestor.
+function getElementsNextToCommonAncestor( positionA, positionB ) {
```

Comment:
This does something different than its name. This method returns elements under common ancestor not next to common ancestor.

`<ancestor><h1>[</h1><h2>]</h2></ancestor><nextToAncestor></nextToAncestor>`

The method will return `<h1>` and `<h2>` not `<nextToAncestor>`.